File: database/app/storage.py
"""Conversation transcript and media-attachment storage on disk.

Large content (the running transcript, attached media) is kept out of the
database and written under settings.media_root, with only the resulting
path stored on the relevant row.
"""

import base64
import secrets
from pathlib import Path

from app.config import settings


def conversation_dir(conversation_id: int) -> Path:
    directory = settings.media_root / str(conversation_id)
    directory.mkdir(parents=True, exist_ok=True)
    return directory


def transcript_path(conversation_id: int) -> Path:
    return conversation_dir(conversation_id) / "transcript.txt"


def append_transcript(conversation_id: int, line: str) -> None:
    with transcript_path(conversation_id).open("a", encoding="utf-8") as f:
        f.write(line + "\n")


def save_media(conversation_id: int, filename: str, base64_data: str) -> str:
    """Decode base64 file data and save it under this conversation's media folder.

    Only the base filename (never any directory components) is used, so a
    caller-supplied name like "../../evil.txt" can't write outside of
    settings.media_root.
    """
    safe_name = f"{secrets.token_hex(4)}_{Path(filename).name}"
    directory = conversation_dir(conversation_id) / "media"
    directory.mkdir(parents=True, exist_ok=True)
    path = directory / safe_name
    path.write_bytes(base64.b64decode(base64_data))
    return str(path)

File: frontend/storage.js
// Conversation transcript + media attachment storage on disk. Mirrors
// ../database/app/storage.py: large content (the running transcript,
// attached media) is written to files under MEDIA_ROOT, with only the
// resulting path stored in the database.
const fs = require("node:fs");
const path = require("node:path");
const crypto = require("node:crypto");

let mediaRoot = process.env.MEDIA_ROOT || path.join(__dirname, "conversation_files");

function setMediaRoot(dir) {
    mediaRoot = dir;
}

function conversationDir(conversationId) {
    const dir = path.join(mediaRoot, String(conversationId));
    fs.mkdirSync(dir, { recursive: true });
    return dir;
}

function transcriptPath(conversationId) {
    return path.join(conversationDir(conversationId), "transcript.txt");
}

function appendTranscript(conversationId, line) {
    fs.appendFileSync(transcriptPath(conversationId), line + "\n", "utf8");
}

// Only the base filename (never any directory components) is used, so a
// caller-supplied name like "../../evil.txt" can't write outside mediaRoot.
function saveMedia(conversationId, filename, base64Data) {
    const safeName = `${crypto.randomBytes(4).toString("hex")}_${path.basename(filename)}`;
    const dir = path.join(conversationDir(conversationId), "media");
    fs.mkdirSync(dir, { recursive: true });
    const filePath = path.join(dir, safeName);
    fs.writeFileSync(filePath, Buffer.from(base64Data, "base64"));
    return filePath;
}

module.exports = { conversationDir, transcriptPath, appendTranscript, saveMedia, setMediaRoot };
